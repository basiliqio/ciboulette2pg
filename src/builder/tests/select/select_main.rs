use super::*;

#[test]
fn all() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url = Url::parse("http://localhost/peoples").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Read;

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &None);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteReadRequest::try_from(request).unwrap();
    let builder =
        Ciboulette2PgBuilder::gen_select(&ciboulette_store, &table_store, &ciboulette_request)
            .unwrap();
    let res = builder.build().unwrap();

    test_sql!(res);
}

#[test]
fn single_id() {
    let ciboulette_store = gen_bag();
    let table_store = gen_table_store(&ciboulette_store);
    let parsed_url =
        Url::parse("http://localhost/peoples/6720877a-e27e-4e9e-9ac0-3fff4deb55f2").unwrap();
    const INTENTION: CibouletteIntention = CibouletteIntention::Read;

    let req_builder = CibouletteRequestBuilder::new(INTENTION, &parsed_url, &None);
    let request = req_builder.build(&ciboulette_store).unwrap();
    let ciboulette_request = CibouletteReadRequest::try_from(request).unwrap();
    let builder =
        Ciboulette2PgBuilder::gen_select(&ciboulette_store, &table_store, &ciboulette_request)
            .unwrap();
    let res = builder.build().unwrap();

    test_sql!(res);
}
