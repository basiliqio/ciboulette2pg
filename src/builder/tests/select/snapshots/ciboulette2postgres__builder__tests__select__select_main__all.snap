---
source: src/builder/tests/select/select_main.rs
expression: "sqlformat::format(query.as_str(), &sqlformat::QueryParams::None,\n                  sqlformat::FormatOptions::default())"

---
WITH "cte_peoples_data" AS (
  SELECT
    "public"."peoples"."id" :: TEXT AS "id",
    $0 :: TEXT AS "type",
    JSON_BUILD_OBJECT(
      $1,
      "public"."peoples"."age",
      $2,
      "public"."peoples"."first-name",
      $3,
      "public"."peoples"."gender",
      $4,
      "public"."peoples"."last-name",
      $5,
      "public"."peoples"."twitter"
    ) AS "data"
  FROM
    "public"."peoples"
),
"cte_favorite_color_data" AS (
  SELECT
    "public"."favorite_color"."id" :: TEXT AS "id",
    $6 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."favorite_color"
  WHERE
    "public"."favorite_color"."id" IN (
      SELECT
        "cte_peoples_data"."favorite_color"
      FROM
        "cte_peoples_data"
    )
),
"cte_rel_articles_rel_data" AS (
  SELECT
    "public"."people-article"."id" :: TEXT AS "id",
    $7 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."people-article"
  WHERE
    "public"."people-article"."people_id" = "cte_peoples_data"."id"
),
"cte_rel_articles_data" AS (
  SELECT
    "public"."articles"."id" :: TEXT AS "id",
    $8 :: TEXT AS "type",
    NULL :: json AS "data"
  FROM
    "public"."articles"
  WHERE
    "public"."articles"."id" IN (
      SELECT
        "cte_rel_articles_rel_data"."article_id"
      FROM
        "cte_rel_articles_rel_data"
    )
)
SELECT
  *
FROM
  "cte_rel_articles_data"
UNION
ALL
SELECT
  *
FROM
  "cte_favorite_color_data"
UNION
ALL
SELECT
  *
FROM
  "cte_rel_articles_rel_data"
UNION
ALL
SELECT
  *
FROM
  "cte_peoples_data";
